core/src/main/java/ai/timefold/solver/core/impl/solver/termination/AbstractCompositeTermination.java

```diff
@@ -3,89 +3,110 @@
 import ai.timefold.solver.core.impl.phase.scope.AbstractStepScope;
 import ai.timefold.solver.core.impl.solver.scope.SolverScope;
 import ai.timefold.solver.core.impl.solver.thread.ChildThreadType;
 
+import org.jspecify.annotations.NullMarked;
+
 /**
  * Abstract superclass that combines multiple {@link Termination}s.
  *
  * @see AndCompositeTermination
  * @see OrCompositeTermination
  */
-public abstract sealed class AbstractCompositeTermination<Solution_>
-        extends AbstractTermination<Solution_>
+@NullMarked
+abstract sealed class AbstractCompositeTermination<Solution_>
+        extends AbstractUniversalTermination<Solution_>
         permits AndCompositeTermination, OrCompositeTermination {
 
     protected final List<Termination<Solution_>> terminationList;
+    protected final List<PhaseTermination<Solution_>> phaseTerminationList;
+    protected final List<SolverTermination<Solution_>> solverTerminationList;
 
     protected AbstractCompositeTermination(List<Termination<Solution_>> terminationList) {
-        this.terminationList = terminationList;
+        this.terminationList = Objects.requireNonNull(terminationList);
+        this.phaseTerminationList = terminationList.stream()
+                .filter(PhaseTermination.class::isInstance)
+                .map(t -> (PhaseTermination<Solution_>) t)
+                .toList();
+        this.solverTerminationList = terminationList.stream()
+                .filter(SolverTermination.class::isInstance)
+                .map(t -> (SolverTermination<Solution_>) t)
+                .toList();
     }
 
+    @SafeVarargs
     public AbstractCompositeTermination(Termination<Solution_>... terminations) {
         this(Arrays.asList(terminations));
     }
 
-    // ************************************************************************
-    // Lifecycle methods
-    // ************************************************************************
-
     @Override
-    public void solvingStarted(SolverScope<Solution_> solverScope) {
-        for (Termination<Solution_> termination : terminationList) {
+    public final void solvingStarted(SolverScope<Solution_> solverScope) {
+        for (var termination : solverTerminationList) {
             termination.solvingStarted(solverScope);
         }
     }
 
     @Override
-    public void phaseStarted(AbstractPhaseScope<Solution_> phaseScope) {
-        for (Termination<Solution_> termination : terminationList) {
+    public final void phaseStarted(AbstractPhaseScope<Solution_> phaseScope) {
+        for (var termination : phaseTerminationList) {
             termination.phaseStarted(phaseScope);
         }
     }
 
     @Override
-    public void stepStarted(AbstractStepScope<Solution_> stepScope) {
-        for (Termination<Solution_> termination : terminationList) {
+    public final void stepStarted(AbstractStepScope<Solution_> stepScope) {
+        for (var termination : phaseTerminationList) {
```

Comment:
The `step` methods do not get called for the solver terminations? How does the solver terminations get updated then?